server/src/main/java/org/elasticsearch/script/ObjectFieldScript.java

```diff
@@ -45,12 +45,13 @@ public ObjectFieldScript(String fieldName, Map<String, Object> params, SearchLoo
      */
     public final List<Object> getValues(String field) {
         //TODO for now we re-run the script every time a leaf field is accessed, but we could cache the values?
+        assert field.startsWith(this.fieldName + ".");
         fieldValues.clear();
         execute();
-        return fieldValues.get(field);
+        return fieldValues.get(field.substring(fieldName.length() + 1));
```

Comment:
I took a slightly different approach: I went for storing values keyed by the full name instead, so no substrings are needed later.